contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,250 @@
+    assert_nn(_voting_period)
+    assert_not_zero(_voting_strategy)
+    assert_not_zero(_authenticator)
+    # TODO: maybe use uint256_signed_nn to check proposal_threshold?
+
+    # Initialize the storage variables
+    voting_delay.write(_voting_delay)
+    voting_period.write(_voting_period)
+    proposal_threshold.write(_proposal_threshold)
+    voting_strategy.write(_voting_strategy)
+    authenticator.write(_authenticator)
+    next_proposal_nonce.write(1)
+
+    return ()
+end
+
+@external
+func vote{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        voter_address : EthAddress, proposal_id : felt, choice : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (proposal) = proposal_registry.read(proposal_id)
+    let (current_block) = get_block_number()
+
+    # Make sure proposal is not closed
+    assert_lt(current_block, proposal.end_block)
+
+    # Make sure proposal has started
+    assert_le(proposal.start_block, current_block)
+
+    let (local params : felt*) = alloc()
+    let (strategy_contract) = voting_strategy.read()
+
+    # TODO: pass in `params_len` and `params`
+    let (voting_power) = IVotingStrategy.get_voting_power(
+        contract_address=strategy_contract,
+        address=voter_address,
+        at=current_block,
+        params_len=0,
+        params=params)
+
+    if choice == Choice.FOR:
+        let (for) = power_for.read(proposal_id)
+        # TODO: what should we do about the carry
+        let (power, carry) = uint256_add(for, voting_power)
+        power_for.write(proposal_id, power)
+        tempvar range_check_ptr = range_check_ptr
+        tempvar syscall_ptr = syscall_ptr
+        tempvar pedersen_ptr = pedersen_ptr
+        tempvar voting_power = voting_power
+    else:
+        if choice == Choice.AGAINST:
+            let (against) = power_against.read(proposal_id)
+            # TODO: what should we do about the carry
+            let (power, carry) = uint256_add(against, voting_power)
+            power_against.write(proposal_id, power)
+            tempvar range_check_ptr = range_check_ptr
+            tempvar syscall_ptr = syscall_ptr
+            tempvar pedersen_ptr = pedersen_ptr
+            tempvar voting_power = voting_power
+        else:
+            if choice == Choice.ABSTAIN:
+                let (_abstain) = power_abstain.read(proposal_id)
+                # TODO: what should we do about the carry
+                let (power, carry) = uint256_add(_abstain, voting_power)
+                power_abstain.write(proposal_id, power)
+                tempvar range_check_ptr = range_check_ptr
+                tempvar syscall_ptr = syscall_ptr
+                tempvar pedersen_ptr = pedersen_ptr
+                tempvar voting_power = voting_power
+            else:
+                # choice is not a valid choice ?!
+                return ()
+            end
+        end
+    end
```

Comment:
We might be able to remove this bad boy if we use this: https://github.com/snapshot-labs/sx-core/pull/42#discussion_r818913667

(and simply  check that `Choice.Against <= choice <= Choice.Abstain`)

Reply:
(implemented on this branch if you want to 👀 it : https://github.com/snapshot-labs/sx-core/commit/7b2060fe034945aa98ea351fe21f4cce7ed229bb)

Reply:
If it work i'm all for it, is there a way to compare the number of steps and amount of storage required?

Reply:
I get the exact same number: `518` steps :)

Reply:
I get the exact same number: `518` steps :)